# Pandas Tutorial
## Intro to DataFrames

In [1]:
import pandas as pd

In [2]:
df = pd.DataFrame([[1,2,3], [4,5,6,],[7,8,9]], columns= ['x','y','z'])

In [3]:
df

x  y  z
0  1  2  3
1  4  5  6
2  7  8  9

In [4]:
df.head()

x  y  z
0  1  2  3
1  4  5  6
2  7  8  9

In [5]:
df.index

RangeIndex(start=0, stop=3, step=1)

In [6]:
df.index = ['a', 'b', 'c']

In [7]:
df

x  y  z
a  1  2  3
b  4  5  6
c  7  8  9

In [10]:
# check the info of memory
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3 entries, a to c
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype
---  ------  --------------  -----
 0   x       3 non-null      int64
 1   y       3 non-null      int64
 2   z       3 non-null      int64
dtypes: int64(3)
memory usage: 96.0+ bytes


In [11]:
df.shape

(3, 3)

In [12]:
# check unique value of a column
df.x.unique()

array([1, 4, 7], dtype=int64)

# Loading in DataFrams From Files
```python
coffee = pd.read_csv('../data/coffe.csv')
coffee = pd.read_parquet('../data/coffe.parquet')
coffee = pd.read_excel('../data/coffe.xlsx')
```

In [16]:
coffee = pd.read_csv('../data/coffe.csv')

In [21]:
# you can save also
coffee.to_csv('../data/coffe_backup.csv')

# Accessing data with pandas

In [18]:
coffee.head()

Day  Coffee Type   Units Sold 
0     Monday     Espresso            25
1     Monday        Latte            15
2    Tuesday     Espresso            30
3    Tuesday        Latte            20
4  Wednesday     Espresso            35

In [19]:
# last five rows
coffee.tail()

Day  Coffee Type   Units Sold 
9     Friday        Latte            35
10  Saturday     Espresso            45
11  Saturday        Latte            35
12    Sunday     Espresso            45
13    Sunday        Latte            35

In [20]:
# random rows
coffee.sample(5)

Day  Coffee Type   Units Sold 
13     Sunday        Latte            35
6    Thursday     Espresso            40
9      Friday        Latte            35
4   Wednesday     Espresso            35
1      Monday        Latte            15

## loc vs iloc

In [22]:
coffee.loc[[0,3,5]]

Day  Coffee Type   Units Sold 
0     Monday     Espresso            25
3    Tuesday        Latte            20
5  Wednesday        Latte            25

In [23]:
coffee.loc[5:8]

Day  Coffee Type   Units Sold 
5  Wednesday        Latte            25
6   Thursday     Espresso            40
7   Thursday        Latte            30
8     Friday     Espresso            45

In [24]:
coffee.loc[5:8, ['Day']]

Day
5  Wednesday
6   Thursday
7   Thursday
8     Friday

In [26]:
coffee.loc[:,['Day']]

Day
0      Monday
1      Monday
2     Tuesday
3     Tuesday
4   Wednesday
5   Wednesday
6    Thursday
7    Thursday
8      Friday
9      Friday
10   Saturday
11   Saturday
12     Sunday
13     Sunday

### iloc

In [28]:
coffee.iloc[:,:2]

Day  Coffee Type
0      Monday     Espresso
1      Monday        Latte
2     Tuesday     Espresso
3     Tuesday        Latte
4   Wednesday     Espresso
5   Wednesday        Latte
6    Thursday     Espresso
7    Thursday        Latte
8      Friday     Espresso
9      Friday        Latte
10   Saturday     Espresso
11   Saturday        Latte
12     Sunday     Espresso
13     Sunday        Latte

In [29]:
coffee.iloc[:5,:5]

Day  Coffee Type   Units Sold 
0     Monday     Espresso            25
1     Monday        Latte            15
2    Tuesday     Espresso            30
3    Tuesday        Latte            20
4  Wednesday     Espresso            35

## Get specific value using `at` or `iat`

In [30]:
coffee.iat[0,0]

'Monday'

In [31]:
# same as this above single value you can use above 
coffee.iloc[0,0]

'Monday'

## Sort the dataframe

In [32]:
coffee.sort_values("Day")

Day  Coffee Type   Units Sold 
8      Friday     Espresso            45
9      Friday        Latte            35
0      Monday     Espresso            25
1      Monday        Latte            15
10   Saturday     Espresso            45
11   Saturday        Latte            35
12     Sunday     Espresso            45
13     Sunday        Latte            35
6    Thursday     Espresso            40
7    Thursday        Latte            30
2     Tuesday     Espresso            30
3     Tuesday        Latte            20
4   Wednesday     Espresso            35
5   Wednesday        Latte            25

In [33]:
coffee.sort_values("Day", ascending = False)

Day  Coffee Type   Units Sold 
4   Wednesday     Espresso            35
5   Wednesday        Latte            25
2     Tuesday     Espresso            30
3     Tuesday        Latte            20
6    Thursday     Espresso            40
7    Thursday        Latte            30
12     Sunday     Espresso            45
13     Sunday        Latte            35
10   Saturday     Espresso            45
11   Saturday        Latte            35
0      Monday     Espresso            25
1      Monday        Latte            15
8      Friday     Espresso            45
9      Friday        Latte            35

In [40]:
coffee.columns

Index(['Day', ' Coffee Type', ' Units Sold '], dtype='object')

In [43]:
# sort by multiple columns
coffee.sort_values(['Day', ' Coffee Type', ' Units Sold '], ascending=False)

Day  Coffee Type   Units Sold 
5   Wednesday        Latte            25
4   Wednesday     Espresso            35
3     Tuesday        Latte            20
2     Tuesday     Espresso            30
7    Thursday        Latte            30
6    Thursday     Espresso            40
13     Sunday        Latte            35
12     Sunday     Espresso            45
11   Saturday        Latte            35
10   Saturday     Espresso            45
1      Monday        Latte            15
0      Monday     Espresso            25
9      Friday        Latte            35
8      Friday     Espresso            45

In [44]:
for index, row in coffee.iterrows():
    print(index, row)

0 Day                Monday
 Coffee Type     Espresso
 Units Sold            25
Name: 0, dtype: object
1 Day             Monday
 Coffee Type     Latte
 Units Sold         15
Name: 1, dtype: object
2 Day               Tuesday
 Coffee Type     Espresso
 Units Sold            30
Name: 2, dtype: object
3 Day             Tuesday
 Coffee Type      Latte
 Units Sold          20
Name: 3, dtype: object
4 Day             Wednesday
 Coffee Type     Espresso
 Units Sold            35
Name: 4, dtype: object
5 Day             Wednesday
 Coffee Type        Latte
 Units Sold            25
Name: 5, dtype: object
6 Day              Thursday
 Coffee Type     Espresso
 Units Sold            40
Name: 6, dtype: object
7 Day             Thursday
 Coffee Type       Latte
 Units Sold           30
Name: 7, dtype: object
8 Day                Friday
 Coffee Type     Espresso
 Units Sold            45
Name: 8, dtype: object
9 Day             Friday
 Coffee Type     Latte
 Units Sold         35
Name: 9, dtype: obje

# Filtering the Data

In [50]:
df = pd.read_csv('../data/chicago_public_schools.csv')

In [51]:
df.head()

School_ID                                     NAME_OF_SCHOOL  \
0     610038                  Abraham Lincoln Elementary School   
1     610281  Adam Clayton Powell Paideia Community Academy ...   
2     610185                Adlai E Stevenson Elementary School   
3     609993                    Agustin Lara Elementary Academy   
4     610513                      Air Force Academy High School   

  Elementary, Middle, or High School         Street_Address     City State  \
0                                 ES        615 W Kemper Pl  Chicago    IL   
1                                 ES  7511 S South Shore Dr  Chicago    IL   
2                                 ES     8010 S Kostner Ave  Chicago    IL   
3                                 ES     4619 S Wolcott Ave  Chicago    IL   
4                                 HS        3630 S Wells St  Chicago    IL   

   ZIP_Code    Phone_Number  \
0     60614  (773) 534-5720   
1     60649  (773) 535-6650   
2     60652  (773) 535-2280   
3     60609  (773) 535-4389   
4     60609  (773) 535-1590   

                                                Link  \
0  http://schoolreports.cps.edu/SchoolProgressRep...   
1  http://schoolreports.cps.edu/SchoolProgressRep...   
2  http://schoolreports.cps.edu/SchoolProgressRep...   
3  http://schoolreports.cps.edu/SchoolProgressRep...   
4  http://schoolreports.cps.edu/SchoolProgressRep...   

                      Network_Manager  ... Freshman_on_Track_Rate__  \
0        Fullerton Elementary Network  ...                      NDA   
1           Skyway Elementary Network  ...                      NDA   
2           Midway Elementary Network  ...                      NDA   
3         Pershing Elementary Network  ...                      NDA   
4  Southwest Side High School Network  ...                      91.   

  X_COORDINATE Y_COORDINATE   Latitude  Longitude COMMUNITY_AREA_NUMBER  \
0  1171699.458  1915829.428  41.924497 -87.644522                     7   
1  1196129.985  1856209.466  41.760324 -87.556736                    43   
2  1148427.165  1851012.215  41.747111 -87.731702                    70   
3  1164504.290  1873959.199  41.809757 -87.672145                    61   
4  1175177.622  1880745.126  41.828146 -87.632794                    34   

  COMMUNITY_AREA_NAME  Ward Police_District                     Location  
0        LINCOLN PARK    43              18  (41.92449696, -87.64452163)  
1         SOUTH SHORE     7               4  (41.76032435, -87.55673627)  
2             ASHBURN    13               8  (41.74711093, -87.73170248)  
3            NEW CITY    20               9    (41.8097569, -87.6721446)  
4       ARMOUR SQUARE    11               9  (41.82814609, -87.63279369)  

[5 rows x 78 columns]

In [52]:
df.loc[df['COMMUNITY_AREA_NUMBER']>50]

School_ID                                     NAME_OF_SCHOOL  \
2       610185                Adlai E Stevenson Elementary School   
3       609993                    Agustin Lara Elementary Academy   
11      609947                 Alexander Graham Elementary School   
17      609788  Alice L Barnard Computer Math & Science Center...   
22      610339                Amos Alonzo Stagg Elementary School   
..         ...                                                ...   
548     610153  William H Ryder Math & Science Specialty Eleme...   
549     610167  William H Seward Communication Arts Academy El...   
550     609698                        William J Bogan High School   
558     609711                  William Rainey Harper High School   
561     610172                William T Sherman Elementary School   

    Elementary, Middle, or High School        Street_Address     City State  \
2                                   ES    8010 S Kostner Ave  Chicago    IL   
3                                   ES    4619 S Wolcott Ave  Chicago    IL   
11                                  ES      4436 S Union Ave  Chicago    IL   
17                                  ES    10354 S Charles St  Chicago    IL   
22                                  ES      7424 S Morgan St  Chicago    IL   
..                                 ...                   ...      ...   ...   
548                                 ES     8716 S Wallace St  Chicago    IL   
549                                 ES  4600 S Hermitage Ave  Chicago    IL   
550                                 HS        3939 W 79th St  Chicago    IL   
558                                 HS        6520 S Wood St  Chicago    IL   
561                                 ES        1000 W 52nd St  Chicago    IL   

     ZIP_Code    Phone_Number  \
2       60652  (773) 535-2280   
3       60609  (773) 535-4389   
11      60609  (773) 535-1308   
17      60643  (773) 535-2625   
22      60621  (773) 535-3565   
..        ...             ...   
548     60620  (773) 535-3843   
549     60609  (773) 535-4890   
550     60652  (773) 535-2180   
558     60636  (773) 535-9150   
561     60609  (773) 535-1757   

                                                  Link  \
2    http://schoolreports.cps.edu/SchoolProgressRep...   
3    http://schoolreports.cps.edu/SchoolProgressRep...   
11   http://schoolreports.cps.edu/SchoolProgressRep...   
17   http://schoolreports.cps.edu/SchoolProgressRep...   
22   http://schoolreports.cps.edu/SchoolProgressRep...   
..                                                 ...   
548  http://schoolreports.cps.edu/SchoolProgressRep...   
549  http://schoolreports.cps.edu/SchoolProgressRep...   
550  http://schoolreports.cps.edu/SchoolProgressRep...   
558  http://schoolreports.cps.edu/SchoolProgressRep...   
561  http://schoolreports.cps.edu/SchoolProgressRep...   

                          Network_Manager  ... Freshman_on_Track_Rate__  \
2               Midway Elementary Network  ...                      NDA   
3             Pershing Elementary Network  ...                      NDA   
11            Pershing Elementary Network  ...                      NDA   
17         Rock Island Elementary Network  ...                      NDA   
22   Englewood-Gresham Elementary Network  ...                      NDA   
..                                    ...  ...                      ...   
548  Englewood-Gresham Elementary Network  ...                      NDA   
549           Pershing Elementary Network  ...                      NDA   
550    Southwest Side High School Network  ...                      67.   
558    Southwest Side High School Network  ...                      64.   
561                          AUSL Schools  ...                      NDA   

    X_COORDINATE Y_COORDINATE   Latitude  Longitude COMMUNITY_AREA_NUMBER  \
2    1148427.165  1851012.215  41.747111 -87.731702                    70   
3    1164504.290  1873959.199  41.809757 -87.672145                    61   
1

In [54]:
# grab some specific columns
df.loc[df['COMMUNITY_AREA_NUMBER']>50, ['COMMUNITY_AREA_NUMBER', 'City']]

COMMUNITY_AREA_NUMBER     City
2                       70  Chicago
3                       61  Chicago
11                      61  Chicago
17                      72  Chicago
22                      68  Chicago
..                     ...      ...
548                     71  Chicago
549                     61  Chicago
550                     70  Chicago
558                     67  Chicago
561                     61  Chicago

[177 rows x 2 columns]

In [56]:
# apply multiple conditions
# grab some specific columns
df.loc[(df['COMMUNITY_AREA_NUMBER']>50) & (df['COMMUNITY_AREA_NAME']=='NEW CITY')]

School_ID                                     NAME_OF_SCHOOL  \
3       609993                    Agustin Lara Elementary Academy   
11      609947                 Alexander Graham Elementary School   
33      610037                   Arthur A Libby Elementary School   
65      610148  Cesar E Chavez Multicultural Academic Center E...   
106     609885              Dewey Elementary Academy of Fine Arts   
130     609735  Edward Tilden Career Community Academy High Sc...   
140     609682        Ellen H Richards Career Academy High School   
246     609983                     James Hedges Elementary School   
283     609964                   John H Hamline Elementary School   
449     610239                 Richard J Daley Elementary Academy   
456     609929                    Robert Fulton Elementary School   
549     610167  William H Seward Communication Arts Academy El...   
561     610172                William T Sherman Elementary School   

    Elementary, Middle, or High School         Street_Address     City State  \
3                                   ES     4619 S Wolcott Ave  Chicago    IL   
11                                  ES       4436 S Union Ave  Chicago    IL   
33                                  ES     5300 S Loomis Blvd  Chicago    IL   
65                                  ES  4747 S Marshfield Ave  Chicago    IL   
106                                 ES       5415 S Union Ave  Chicago    IL   
130                                 HS       4747 S Union Ave  Chicago    IL   
140                                 HS       5009 S Laflin St  Chicago    IL   
246                                 ES  4747 S Winchester Ave  Chicago    IL   
283                                 ES       4747 S Bishop St  Chicago    IL   
449                                 ES     5024 S Wolcott Ave  Chicago    IL   
456                                 ES   5300 S Hermitage Ave  Chicago    IL   
549                                 ES   4600 S Hermitage Ave  Chicago    IL   
561                                 ES         1000 W 52nd St  Chicago    IL   

     ZIP_Code    Phone_Number  \
3       60609  (773) 535-4389   
11      60609  (773) 535-1308   
33      60609  (773) 535-9050   
65      60609  (773) 535-4600   
106     60609  (773) 535-1666   
130     60609  (773) 535-1625   
140     60609  (773) 535-4945   
246     60609  (773) 535-7360   
283     60609  (773) 535-4565   
449     60609  (773) 535-9091   
456     60609  (773) 535-9000   
549     60609  (773) 535-4890   
561     60609  (773) 535-1757   

                                                  Link  \
3    http://schoolreports.cps.edu/SchoolProgressRep...   
11   http://schoolreports.cps.edu/SchoolProgressRep...   
33   http://schoolreports.cps.edu/SchoolProgressRep...   
65   http://schoolreports.cps.edu/SchoolProgressRep...   
106  http://schoolreports.cps.edu/SchoolProgressRep...   
130  http://schoolreports.cps.edu/SchoolProgressRep...   
140  http://schoolreports.cps.edu/SchoolProgressRep...   
246  http://schoolreports.cps.edu/SchoolProgressRep...   
283  http://schoolreports.cps.edu/SchoolProgressRep...   
449  http://schoolreports.cps.edu/SchoolProgressRep...   
456  http://schoolreports.cps.edu/SchoolProgressRep...   
549  http://schoolreports.cps.edu/SchoolProgressRep...   
561  http://schoolreports.cps.edu/SchoolProgressRep...   

                        Network_Manager  ... Freshman_on_Track_Rate__  \
3           Pershing Elementary Network  ...                      NDA   
11          Pershing Elementary Network  ...                      NDA   
33          Pershing Elementary Network  ...                      NDA   
65          Pershing Elementary Network  ...                      NDA   
106         Pershing Elementary Network  ...                      NDA   
130  Southwest Side High School Network  ...                      86.   
140  Southwest Side High School Network  ...                      63.   
246         Pershing Elementary Network  ...             

## Regular Expression 

In [58]:
# search data that contains specific characters
df[df['NAME_OF_SCHOOL'].str.contains('High School')]

School_ID                                     NAME_OF_SCHOOL  \
4       610513                      Air Force Academy High School   
6       609720                Albert G Lane Technical High School   
8       610524              Alcott High School for the Humanities   
18      610334                                Al Raby High School   
40      400018  Austin Business and Entrepreneurship Academy H...   
..         ...                                                ...   
545     609734                    William Howard Taft High School   
550     609698                        William J Bogan High School   
553     609678      William Jones College Preparatory High School   
558     609711                  William Rainey Harper High School   
565     610392                 World Language Academy High School   

    Elementary, Middle, or High School        Street_Address     City State  \
4                                   HS       3630 S Wells St  Chicago    IL   
6                                   HS     2501 W Addison St  Chicago    IL   
8                                   HS      2957 N Hoyne Ave  Chicago    IL   
18                                  HS    3545 W Fulton Blvd  Chicago    IL   
40                                  HS        231 N Pine Ave  Chicago    IL   
..                                 ...                   ...      ...   ...   
545                                 HS  6530 W Bryn Mawr Ave  Chicago    IL   
550                                 HS        3939 W 79th St  Chicago    IL   
553                                 HS        606 S State St  Chicago    IL   
558                                 HS        6520 S Wood St  Chicago    IL   
565                                 HS    3120 S Kostner Ave  Chicago    IL   

     ZIP_Code    Phone_Number  \
4       60609  (773) 535-1590   
6       60618  (773) 534-5400   
8       60618  (773) 534-5979   
18      60624  (773) 534-6755   
40      60644  (773) 534-6316   
..        ...             ...   
545     60631  (773) 534-1000   
550     60652  (773) 535-2180   
553     60605  (773) 534-8600   
558     60636  (773) 535-9150   
565     60623  (773) 535-4334   

                                                  Link  \
4    http://schoolreports.cps.edu/SchoolProgressRep...   
6    http://schoolreports.cps.edu/SchoolProgressRep...   
8    http://schoolreports.cps.edu/SchoolProgressRep...   
18   http://schoolreports.cps.edu/SchoolProgressRep...   
40                                                 NaN   
..                                                 ...   
545  http://schoolreports.cps.edu/SchoolProgressRep...   
550  http://schoolreports.cps.edu/SchoolProgressRep...   
553  http://schoolreports.cps.edu/SchoolProgressRep...   
558  http://schoolreports.cps.edu/SchoolProgressRep...   
565  http://schoolreports.cps.edu/SchoolProgressRep...   

                              Network_Manager  ... Freshman_on_Track_Rate__  \
4          Southwest Side High School Network  ...                      91.   
6    North-Northwest Side High School Network  ...                      90.   
8    North-Northwest Side High School Network  ...                      87.   
18              West Side High School Network  ...                      74.   
40              West Side High School Network  ...                      NDA   
..                                        ...  ...                      ...   
545  North-Northwest Side High School Network  ...                      77.   
550        Southwest Side High School Network  ...                      67.   
553            South Side High School Network  ...                      92.   
558        Southwest Side High School Network  ...                      64.   
565             West Side High School Network  ...                       76   

    X_COORDINATE Y_COORDINATE   Latitude  Longitude COMMUNITY_AREA_NUMBER  \
4    1175177.622  1880745.126  41.828146 -87.632794                    34   
6    1158975.392  1923791.705 

In [61]:
df[df['NAME_OF_SCHOOL'].str.contains('High School | College')]

School_ID                                     NAME_OF_SCHOOL  \
8       610524              Alcott High School for the Humanities   
83      609753      Chicago High School for Agricultural Sciences   
112     609751  Dr  Martin Luther King  Jr  College Prep High ...   
203     610383    Greater Lawndale High School For Social Justice   
207     609726  Gwendolyn Brooks College Preparatory Academy H...   
222     610297           High School of Leadership at South Shore   
224     609768               Hope College Preparatory High School   
280     609694       John Hancock College Preparatory High School   
400     610323      New Millennium High School of Health at Bowen   
408     609749          Northside College Preparatory High School   
518     609680      Walter Payton College Preparatory High School   
553     609678      William Jones College Preparatory High School   

    Elementary, Middle, or High School        Street_Address     City State  \
8                                   HS      2957 N Hoyne Ave  Chicago    IL   
83                                  HS       3857 W 111th St  Chicago    IL   
112                                 HS    4445 S Drexel Blvd  Chicago    IL   
203                                 HS    3120 S Kostner Ave  Chicago    IL   
207                                 HS        250 E 111th St  Chicago    IL   
222                                 HS  7627 S Constance Ave  Chicago    IL   
224                                 HS       5515 S Lowe Ave  Chicago    IL   
280                                 HS        4034 W 56th St  Chicago    IL   
400                                 HS        2710 E 89th St  Chicago    IL   
408                                 HS     5501 N Kedzie Ave  Chicago    IL   
518                                 HS       1034 N Wells St  Chicago    IL   
553                                 HS        606 S State St  Chicago    IL   

     ZIP_Code    Phone_Number  \
8       60618  (773) 534-5979   
83      60655  (773) 535-2500   
112     60653  (773) 535-1180   
203     60623  (773) 535-4300   
207     60628  (773) 535-9930   
222     60649  (773) 535-6190   
224     60621  (773) 535-3160   
280     60629  (773) 535-2410   
400     60617  (773) 535-7650   
408     60625  (773) 534-3954   
518     60610  (773) 534-0034   
553     60605  (773) 534-8600   

                                                  Link  \
8    http://schoolreports.cps.edu/SchoolProgressRep...   
83   http://schoolreports.cps.edu/SchoolProgressRep...   
112  http://schoolreports.cps.edu/SchoolProgressRep...   
203  http://schoolreports.cps.edu/SchoolProgressRep...   
207  http://schoolreports.cps.edu/SchoolProgressRep...   
222  http://schoolreports.cps.edu/SchoolProgressRep...   
224  http://schoolreports.cps.edu/SchoolProgressRep...   
280  http://schoolreports.cps.edu/SchoolProgressRep...   
400  http://schoolreports.cps.edu/SchoolProgressRep...   
408  http://schoolreports.cps.edu/SchoolProgressRep...   
518  http://schoolreports.cps.edu/SchoolProgressRep...   
553  http://schoolreports.cps.edu/SchoolProgressRep...   

                              Network_Manager  ... Freshman_on_Track_Rate__  \
8    North-Northwest Side High School Network  ...                      87.   
83         Far South Side High School Network  ...                      88.   
112            South Side High School Network  ...                      91.   
203             West Side High School Network  ...                      87.   
207        Far South Side High School Network  ...                      90.   
222            South Side High School Network  ...                      48.   
224        Southwest Side High School Network  ...                      63.   
280        Southwest Side High School Network  ...                      87.   
400            South Side High School Network  ...                      62.   
408  North-Northwest Side High School Network  ...                      95.   
518  North-Northwest Side

In [66]:
# find with regex the vowel area
df[df['COMMUNITY_AREA_NAME'].str.contains(r'^[AIEOUaeiou]')]['COMMUNITY_AREA_NAME']

2                 ASHBURN
4           ARMOUR SQUARE
5             ALBANY PARK
9             ALBANY PARK
18     EAST GARFIELD PARK
              ...        
535               ASHBURN
537                UPTOWN
542           ALBANY PARK
548        AUBURN GRESHAM
550               ASHBURN
Name: COMMUNITY_AREA_NAME, Length: 122, dtype: object

In [72]:
df[df['COMMUNITY_AREA_NAME'].isin(['NEW CITY'])]

School_ID                                     NAME_OF_SCHOOL  \
3       609993                    Agustin Lara Elementary Academy   
11      609947                 Alexander Graham Elementary School   
33      610037                   Arthur A Libby Elementary School   
65      610148  Cesar E Chavez Multicultural Academic Center E...   
106     609885              Dewey Elementary Academy of Fine Arts   
130     609735  Edward Tilden Career Community Academy High Sc...   
140     609682        Ellen H Richards Career Academy High School   
246     609983                     James Hedges Elementary School   
283     609964                   John H Hamline Elementary School   
449     610239                 Richard J Daley Elementary Academy   
456     609929                    Robert Fulton Elementary School   
549     610167  William H Seward Communication Arts Academy El...   
561     610172                William T Sherman Elementary School   

    Elementary, Middle, or High School         Street_Address     City State  \
3                                   ES     4619 S Wolcott Ave  Chicago    IL   
11                                  ES       4436 S Union Ave  Chicago    IL   
33                                  ES     5300 S Loomis Blvd  Chicago    IL   
65                                  ES  4747 S Marshfield Ave  Chicago    IL   
106                                 ES       5415 S Union Ave  Chicago    IL   
130                                 HS       4747 S Union Ave  Chicago    IL   
140                                 HS       5009 S Laflin St  Chicago    IL   
246                                 ES  4747 S Winchester Ave  Chicago    IL   
283                                 ES       4747 S Bishop St  Chicago    IL   
449                                 ES     5024 S Wolcott Ave  Chicago    IL   
456                                 ES   5300 S Hermitage Ave  Chicago    IL   
549                                 ES   4600 S Hermitage Ave  Chicago    IL   
561                                 ES         1000 W 52nd St  Chicago    IL   

     ZIP_Code    Phone_Number  \
3       60609  (773) 535-4389   
11      60609  (773) 535-1308   
33      60609  (773) 535-9050   
65      60609  (773) 535-4600   
106     60609  (773) 535-1666   
130     60609  (773) 535-1625   
140     60609  (773) 535-4945   
246     60609  (773) 535-7360   
283     60609  (773) 535-4565   
449     60609  (773) 535-9091   
456     60609  (773) 535-9000   
549     60609  (773) 535-4890   
561     60609  (773) 535-1757   

                                                  Link  \
3    http://schoolreports.cps.edu/SchoolProgressRep...   
11   http://schoolreports.cps.edu/SchoolProgressRep...   
33   http://schoolreports.cps.edu/SchoolProgressRep...   
65   http://schoolreports.cps.edu/SchoolProgressRep...   
106  http://schoolreports.cps.edu/SchoolProgressRep...   
130  http://schoolreports.cps.edu/SchoolProgressRep...   
140  http://schoolreports.cps.edu/SchoolProgressRep...   
246  http://schoolreports.cps.edu/SchoolProgressRep...   
283  http://schoolreports.cps.edu/SchoolProgressRep...   
449  http://schoolreports.cps.edu/SchoolProgressRep...   
456  http://schoolreports.cps.edu/SchoolProgressRep...   
549  http://schoolreports.cps.edu/SchoolProgressRep...   
561  http://schoolreports.cps.edu/SchoolProgressRep...   

                        Network_Manager  ... Freshman_on_Track_Rate__  \
3           Pershing Elementary Network  ...                      NDA   
11          Pershing Elementary Network  ...                      NDA   
33          Pershing Elementary Network  ...                      NDA   
65          Pershing Elementary Network  ...                      NDA   
106         Pershing Elementary Network  ...                      NDA   
130  Southwest Side High School Network  ...                      86.   
140  Southwest Side High School Network  ...                      63.   
246         Pershing Elementary Network  ...             

### Another way of query data

In [75]:
df.query("COMMUNITY_AREA_NAME == 'NEW CITY' and Ward >16")

School_ID                                     NAME_OF_SCHOOL  \
3       609993                    Agustin Lara Elementary Academy   
65      610148  Cesar E Chavez Multicultural Academic Center E...   
246     609983                     James Hedges Elementary School   
283     609964                   John H Hamline Elementary School   
549     610167  William H Seward Communication Arts Academy El...   

    Elementary, Middle, or High School         Street_Address     City State  \
3                                   ES     4619 S Wolcott Ave  Chicago    IL   
65                                  ES  4747 S Marshfield Ave  Chicago    IL   
246                                 ES  4747 S Winchester Ave  Chicago    IL   
283                                 ES       4747 S Bishop St  Chicago    IL   
549                                 ES   4600 S Hermitage Ave  Chicago    IL   

     ZIP_Code    Phone_Number  \
3       60609  (773) 535-4389   
65      60609  (773) 535-4600   
246     60609  (773) 535-7360   
283     60609  (773) 535-4565   
549     60609  (773) 535-4890   

                                                  Link  \
3    http://schoolreports.cps.edu/SchoolProgressRep...   
65   http://schoolreports.cps.edu/SchoolProgressRep...   
246  http://schoolreports.cps.edu/SchoolProgressRep...   
283  http://schoolreports.cps.edu/SchoolProgressRep...   
549  http://schoolreports.cps.edu/SchoolProgressRep...   

                 Network_Manager  ... Freshman_on_Track_Rate__ X_COORDINATE  \
3    Pershing Elementary Network  ...                      NDA  1164504.290   
65   Pershing Elementary Network  ...                      NDA  1166186.662   
246  Pershing Elementary Network  ...                      NDA  1164199.040   
283  Pershing Elementary Network  ...                      NDA  1167511.897   
549  Pershing Elementary Network  ...                      NDA  1165412.465   

    Y_COORDINATE   Latitude  Longitude COMMUNITY_AREA_NUMBER  \
3    1873959.199  41.809757 -87.672145                    61   
65   1873024.079  41.807155 -87.666001                    61   
246  1873001.047  41.807134 -87.673291                    61   
283  1873073.883  41.807263 -87.661139                    61   
549  1874162.777  41.810296 -87.668808                    61   

    COMMUNITY_AREA_NAME  Ward Police_District                     Location  
3              NEW CITY    20               9    (41.8097569, -87.6721446)  
65             NEW CITY    20               9  (41.80715514, -87.66600055)  
246            NEW CITY    20               9  (41.80713406, -87.67329121)  
283            NEW CITY    20               9  (41.80726347, -87.66113856)  
549            NEW CITY    20               9  (41.81029632, -87.66880778)  

[5 rows x 78 columns]

# Adding and Removing Columns

In [81]:
coffee = pd.read_csv('../data/coffe.csv')

In [82]:
coffee['Price'] = 4.99

In [83]:
coffee.head()

Day Coffee Type  Units Sold  Price
0     Monday    Espresso          25   4.99
1     Monday       Latte          15   4.99
2    Tuesday    Espresso          30   4.99
3    Tuesday       Latte          20   4.99
4  Wednesday    Espresso          35   4.99

In [90]:
import numpy as np
# Assign new_price based on the Coffee Type
coffee['new_price'] = np.where(coffee['Coffee Type'] == ' Latte', 3, 5)

In [92]:
coffee.head()

Day Coffee Type  Units Sold  Price  new_price
0     Monday    Espresso          25   4.99          5
1     Monday       Latte          15   4.99          3
2    Tuesday    Espresso          30   4.99          5
3    Tuesday       Latte          20   4.99          3
4  Wednesday    Espresso          35   4.99          5

In [93]:
# Drop a particular column
coffee.drop(columns=['Price'])

Day Coffee Type  Units Sold  new_price
0      Monday    Espresso          25          5
1      Monday       Latte          15          3
2     Tuesday    Espresso          30          5
3     Tuesday       Latte          20          3
4   Wednesday    Espresso          35          5
5   Wednesday       Latte          25          3
6    Thursday    Espresso          40          5
7    Thursday       Latte          30          3
8      Friday    Espresso          45          5
9      Friday       Latte          35          3
10   Saturday    Espresso          45          5
11   Saturday       Latte          35          3
12     Sunday    Espresso          45          5
13     Sunday       Latte          35          3

In [94]:
# copy the dataframe
coffee_new = coffee.copy()

In [95]:
# to replace the previous dataframe with current changes
coffee.drop(columns=['Price'], inplace=True)

In [96]:
coffee.head()

Day Coffee Type  Units Sold  new_price
0     Monday    Espresso          25          5
1     Monday       Latte          15          3
2    Tuesday    Espresso          30          5
3    Tuesday       Latte          20          3
4  Wednesday    Espresso          35          5

In [100]:
# select only specific columns
coffee = coffee[['Day', 'Coffee Type', 'new_price', 'Units Sold']]

In [101]:
coffee.head()

Day Coffee Type  new_price  Units Sold
0     Monday    Espresso          5          25
1     Monday       Latte          3          15
2    Tuesday    Espresso          5          30
3    Tuesday       Latte          3          20
4  Wednesday    Espresso          5          35

In [102]:
# apply some operation between columns
coffee['revenue'] = coffee['Units Sold'] * coffee['new_price']

In [103]:
coffee.head()

Day Coffee Type  new_price  Units Sold  revenue
0     Monday    Espresso          5          25      125
1     Monday       Latte          3          15       45
2    Tuesday    Espresso          5          30      150
3    Tuesday       Latte          3          20       60
4  Wednesday    Espresso          5          35      175

In [105]:
# Rename a column
coffee.rename(columns={'revenue':'collection'}, inplace= True)

In [106]:
coffee.head()

Day Coffee Type  new_price  Units Sold  collection
0     Monday    Espresso          5          25         125
1     Monday       Latte          3          15          45
2    Tuesday    Espresso          5          30         150
3    Tuesday       Latte          3          20          60
4  Wednesday    Espresso          5          35         175

In [110]:
coffee['Date'] = date_list = [
    "2023-01-01",
    "2023-01-02",
    "2023-01-03",
    "2023-01-04",
    "2023-01-05",
    "2023-01-06",
    "2023-01-07",
    "2023-01-08",
    "2023-01-09",
    "2023-01-10",
    "2023-01-11",
    "2023-01-12",
    "2023-01-13",
    "2023-01-14"
]

In [111]:
coffee.head()

Day Coffee Type  new_price  Units Sold  collection        Date
0     Monday    Espresso          5          25         125  2023-01-01
1     Monday       Latte          3          15          45  2023-01-02
2    Tuesday    Espresso          5          30         150  2023-01-03
3    Tuesday       Latte          3          20          60  2023-01-04
4  Wednesday    Espresso          5          35         175  2023-01-05

In [113]:
# convert the date column to pandas date type
coffee['Date'] = pd.to_datetime(coffee['Date'])

In [114]:
coffee.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 6 columns):
 #   Column       Non-Null Count  Dtype         
---  ------       --------------  -----         
 0   Day          14 non-null     object        
 1   Coffee Type  14 non-null     object        
 2   new_price    14 non-null     int32         
 3   Units Sold   14 non-null     int64         
 4   collection   14 non-null     int64         
 5   Date         14 non-null     datetime64[ns]
dtypes: datetime64[ns](1), int32(1), int64(2), object(2)
memory usage: 748.0+ bytes


In [117]:
# Now you can extract the year date month since it is pandas datetime type
coffee['Year'] = coffee['Date'].dt.year
coffee.head()

Day Coffee Type  new_price  Units Sold  collection       Date  Year
0     Monday    Espresso          5          25         125 2023-01-01  2023
1     Monday       Latte          3          15          45 2023-01-02  2023
2    Tuesday    Espresso          5          30         150 2023-01-03  2023
3    Tuesday       Latte          3          20          60 2023-01-04  2023
4  Wednesday    Espresso          5          35         175 2023-01-05  2023

In [121]:
coffee['Month'] = coffee['Date'].dt.month
coffee.head()

Day Coffee Type  new_price  Units Sold  collection       Date  Year  \
0     Monday    Espresso          5          25         125 2023-01-01  2023   
1     Monday       Latte          3          15          45 2023-01-02  2023   
2    Tuesday    Espresso          5          30         150 2023-01-03  2023   
3    Tuesday       Latte          3          20          60 2023-01-04  2023   
4  Wednesday    Espresso          5          35         175 2023-01-05  2023   

   Month  
0      1  
1      1  
2      1  
3      1  
4      1

In [122]:
coffee.to_csv('../data/new_coffee.csv', index=False)

## `apply` method

### 1st method

In [124]:
coffee['Rating'] = coffee['Units Sold'].apply(lambda x: 'Below Average' if x < 20 else ('Average' if  x < 30 else 'Above Average'))

In [125]:
coffee

Day Coffee Type  new_price  Units Sold  collection       Date  Year  \
0      Monday    Espresso          5          25         125 2023-01-01  2023   
1      Monday       Latte          3          15          45 2023-01-02  2023   
2     Tuesday    Espresso          5          30         150 2023-01-03  2023   
3     Tuesday       Latte          3          20          60 2023-01-04  2023   
4   Wednesday    Espresso          5          35         175 2023-01-05  2023   
5   Wednesday       Latte          3          25          75 2023-01-06  2023   
6    Thursday    Espresso          5          40         200 2023-01-07  2023   
7    Thursday       Latte          3          30          90 2023-01-08  2023   
8      Friday    Espresso          5          45         225 2023-01-09  2023   
9      Friday       Latte          3          35         105 2023-01-10  2023   
10   Saturday    Espresso          5          45         225 2023-01-11  2023   
11   Saturday       Latte          3          35         105 2023-01-12  2023   
12     Sunday    Espresso          5          45         225 2023-01-13  2023   
13     Sunday       Latte          3          35         105 2023-01-14  2023   

    Month         Rating  
0       1        Average  
1       1  Below Average  
2       1  Above Average  
3       1        Average  
4       1  Above Average  
5       1        Average  
6       1  Above Average  
7       1  Above Average  
8       1  Above Average  
9       1  Above Average  
10      1  Above Average  
11      1  Above Average  
12      1  Above Average  
13      1  Above Average

### 2nd method

In [126]:
def categorize_price(row):
    if row >4:
        return 'costly'
    else:
        return 'fair'
coffee['Category'] = coffee['new_price'].apply(categorize_price)
coffee.head()

Day Coffee Type  new_price  Units Sold  collection       Date  Year  \
0     Monday    Espresso          5          25         125 2023-01-01  2023   
1     Monday       Latte          3          15          45 2023-01-02  2023   
2    Tuesday    Espresso          5          30         150 2023-01-03  2023   
3    Tuesday       Latte          3          20          60 2023-01-04  2023   
4  Wednesday    Espresso          5          35         175 2023-01-05  2023   

   Month         Rating Category  
0      1        Average   costly  
1      1  Below Average     fair  
2      1  Above Average   costly  
3      1        Average     fair  
4      1  Above Average   costly

# Merging and Concatening DataFrames

In [1]:
import pandas as pd

In [18]:
df1 = pd.read_csv("../data/employees.csv")

In [19]:
df1.head()

EmployeeID     Name   Department
0       68319    Alice           HR
1       66928      Bob  Engineering
2       67832  Charlie      Finance
3       65646    David    Marketing
4       67858      Eve  Engineering

In [20]:
df1.shape

(9, 3)

In [32]:
df2= pd.read_csv("../data/salary.csv")

In [33]:
df2.head()

EmployeeID  Salary
0       69062   70000
1       63679   80000
2       64989   90000
3       65271   60000
4       66564   95000

In [34]:
df2.shape

(9, 2)

### Merging df1 and df2 on 'EmployeeID' using a right join to include all records from df2

In [45]:
new_df = pd.merge(df1, df2, on='EmployeeID', how='right')
new_df

EmployeeID    Name Department  Salary
0       69062   Frank         HR   70000
1       63679   Grace    Finance   80000
2       64989  Hannah  Marketing   90000
3       65271     Ian         HR   60000
4       66564     NaN        NaN   95000
5       68454     NaN        NaN   76000
6       68736     NaN        NaN   72000
7       69000     NaN        NaN   85000
8       69324     NaN        NaN   50000

### Merging df1 and df2 on 'EmployeeID' using a left join to include all records from df1

In [46]:
new_df = pd.merge(df1, df2, on='EmployeeID', how='left')
new_df 

EmployeeID     Name   Department   Salary
0       68319    Alice           HR      NaN
1       66928      Bob  Engineering      NaN
2       67832  Charlie      Finance      NaN
3       65646    David    Marketing      NaN
4       67858      Eve  Engineering      NaN
5       69062    Frank           HR  70000.0
6       63679    Grace      Finance  80000.0
7       64989   Hannah    Marketing  90000.0
8       65271      Ian           HR  60000.0

### Merging df1 and df2 on 'EmployeeID' using an inner join to include only matching records

In [48]:
new_df = pd.merge(df1, df2, on='EmployeeID', how='inner')
new_df 

EmployeeID    Name Department  Salary
0       69062   Frank         HR   70000
1       63679   Grace    Finance   80000
2       64989  Hannah  Marketing   90000
3       65271     Ian         HR   60000

### Merging df1 and df2 on 'EmployeeID' using the default inner join (same as above)

In [49]:
new_df = pd.merge(df1, df2, on='EmployeeID')
new_df  

EmployeeID    Name Department  Salary
0       69062   Frank         HR   70000
1       63679   Grace    Finance   80000
2       64989  Hannah  Marketing   90000
3       65271     Ian         HR   60000

### Merging df1 and df2 again using a left join for further analysis

In [50]:
new_df = pd.merge(df1, df2, on='EmployeeID', how='left')
new_df  

EmployeeID     Name   Department   Salary
0       68319    Alice           HR      NaN
1       66928      Bob  Engineering      NaN
2       67832  Charlie      Finance      NaN
3       65646    David    Marketing      NaN
4       67858      Eve  Engineering      NaN
5       69062    Frank           HR  70000.0
6       63679    Grace      Finance  80000.0
7       64989   Hannah    Marketing  90000.0
8       65271      Ian           HR  60000.0

In [51]:
# Storing the first five rows of new_df in 'head' variable
head = new_df.head()
# Storing the last five rows of new_df in 'tail' variable
tail = new_df.tail()

### Concatenating the last five rows (tail) with the first five rows (head) into a new DataFrame

In [52]:
concat_df = pd.concat([tail, head])
concat_df  

EmployeeID     Name   Department   Salary
4       67858      Eve  Engineering      NaN
5       69062    Frank           HR  70000.0
6       63679    Grace      Finance  80000.0
7       64989   Hannah    Marketing  90000.0
8       65271      Ian           HR  60000.0
0       68319    Alice           HR      NaN
1       66928      Bob  Engineering      NaN
2       67832  Charlie      Finance      NaN
3       65646    David    Marketing      NaN
4       67858      Eve  Engineering      NaN

#### Filtering new_df to find records where Salary is less than EmployeeID

In [53]:
new_df[new_df['Salary'] < new_df['EmployeeID']]

EmployeeID Name Department   Salary
8       65271  Ian         HR  60000.0

# Handling Null Values

In [57]:
# suppose our dataframe
df = concat_df

In [58]:
df

EmployeeID     Name   Department   Salary
4       67858      Eve  Engineering      NaN
5       69062    Frank           HR  70000.0
6       63679    Grace      Finance  80000.0
7       64989   Hannah    Marketing  90000.0
8       65271      Ian           HR  60000.0
0       68319    Alice           HR      NaN
1       66928      Bob  Engineering      NaN
2       67832  Charlie      Finance      NaN
3       65646    David    Marketing      NaN
4       67858      Eve  Engineering      NaN

In [60]:
# check for non-null values
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 10 entries, 4 to 4
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   EmployeeID  10 non-null     int64  
 1   Name        10 non-null     object 
 2   Department  10 non-null     object 
 3   Salary      4 non-null      float64
dtypes: float64(1), int64(1), object(2)
memory usage: 400.0+ bytes


In [63]:
# check how many null values a column has
df.isna().sum()

EmployeeID    0
Name          0
Department    0
Salary        6
dtype: int64

#### Fill null values

In [66]:
# fill by any value
df.fillna(9999)

EmployeeID     Name   Department   Salary
4       67858      Eve  Engineering   9999.0
5       69062    Frank           HR  70000.0
6       63679    Grace      Finance  80000.0
7       64989   Hannah    Marketing  90000.0
8       65271      Ian           HR  60000.0
0       68319    Alice           HR   9999.0
1       66928      Bob  Engineering   9999.0
2       67832  Charlie      Finance   9999.0
3       65646    David    Marketing   9999.0
4       67858      Eve  Engineering   9999.0

In [71]:
# fill by a mean value
df.fillna(df['Salary'].mean())

EmployeeID     Name   Department   Salary
4       67858      Eve  Engineering  75000.0
5       69062    Frank           HR  70000.0
6       63679    Grace      Finance  80000.0
7       64989   Hannah    Marketing  90000.0
8       65271      Ian           HR  60000.0
0       68319    Alice           HR  75000.0
1       66928      Bob  Engineering  75000.0
2       67832  Charlie      Finance  75000.0
3       65646    David    Marketing  75000.0
4       67858      Eve  Engineering  75000.0

## Dropping duplicate rows

In [96]:
df.drop_duplicates(inplace=True)
df

EmployeeID     Name   Department   Salary
5       69062    Frank           HR  70000.0
6       63679    Grace      Finance  80000.0
7       64989   Hannah    Marketing  90000.0
8       65271      Ian           HR  60000.0
0       68319    Alice           HR      NaN
1       66928      Bob  Engineering      NaN
2       67832  Charlie      Finance      NaN
3       65646    David    Marketing      NaN

### now fill by interpolate

In [97]:
# fill by interpolate
df.fillna({'Salary': df['Salary'].interpolate()})

EmployeeID     Name   Department   Salary
5       69062    Frank           HR  70000.0
6       63679    Grace      Finance  80000.0
7       64989   Hannah    Marketing  90000.0
8       65271      Ian           HR  60000.0
0       68319    Alice           HR  60000.0
1       66928      Bob  Engineering  60000.0
2       67832  Charlie      Finance  60000.0
3       65646    David    Marketing  60000.0

In [98]:
df[df['Salary'].isna()]

EmployeeID     Name   Department  Salary
0       68319    Alice           HR     NaN
1       66928      Bob  Engineering     NaN
2       67832  Charlie      Finance     NaN
3       65646    David    Marketing     NaN

In [99]:
df[df['Salary'].notna()]

EmployeeID    Name Department   Salary
5       69062   Frank         HR  70000.0
6       63679   Grace    Finance  80000.0
7       64989  Hannah  Marketing  90000.0
8       65271     Ian         HR  60000.0

# Aggegating Data
- **Aggregation functions:** Pandas provides built-in functions like sum(), mean(), min(), max(), and count() to compute summary statistics across a DataFrame or Series
- **Grouping:** The groupby() function enables you to split data into groups based on certain criteria before applying aggregation functions, which is particularly useful for analyzing data patterns within categories

In [121]:
import pandas as pd
coffee = pd.read_csv("../data/coffee.csv")

In [122]:
coffee.groupby(['Coffee Type'])['Units Sold'].sum()

Coffee Type
Espresso    265
Latte       195
Name: Units Sold, dtype: int64

In [123]:
coffee.groupby(['Coffee Type'])['Units Sold'].mean()

Coffee Type
Espresso    37.857143
Latte       27.857143
Name: Units Sold, dtype: float64

### Using the agg() function to apply multiple aggregation functions:
- Calculate the mean of 'Units Sold' for each coffee type
 - Calculate the total sum of 'Price' for each coffee type

In [126]:
coffee.groupby(['Coffee Type']).agg({
    'Units Sold': 'mean',  # Calculate the average units sold per coffee type
    'Price': 'sum'         # Calculate the total price for all sales of each coffee type
    })

Units Sold  Price
Coffee Type                   
Espresso      37.857143     35
Latte         27.857143     21

In [127]:
coffee.groupby(['Coffee Type', 'Day']).agg({
    'Units Sold': 'mean',  # Calculate the average units sold per coffee type
    'Price': 'sum'         # Calculate the total price for all sales of each coffee type
    })

Units Sold  Price
Coffee Type Day                         
Espresso    Friday           45.0      5
            Monday           25.0      5
            Saturday         45.0      5
            Sunday           45.0      5
            Thursday         40.0      5
            Tuesday          30.0      5
            Wednesday        35.0      5
Latte       Friday           35.0      3
            Monday           15.0      3
            Saturday         35.0      3
            Sunday           35.0      3
            Thursday         30.0      3
            Tuesday          20.0      3
            Wednesday        25.0      3

### Pivot Tables

In [129]:
coffee.head()

Day Coffee Type  Units Sold  Price
0     Monday    Espresso          25      5
1     Monday       Latte          15      3
2    Tuesday    Espresso          30      5
3    Tuesday       Latte          20      3
4  Wednesday    Espresso          35      5

In [132]:
pivot = coffee.pivot(columns= 'Coffee Type', index='Day', values='Units Sold')
pivot

Coffee Type  Espresso  Latte
Day                         
Friday             45     35
Monday             25     15
Saturday           45     35
Sunday             45     35
Thursday           40     30
Tuesday            30     20
Wednesday          35     25

In [133]:
pivot.sum()

Coffee Type
Espresso    265
Latte       195
dtype: int64

In [134]:
pivot.sum(axis=1)

Day
Friday       80
Monday       40
Saturday     80
Sunday       80
Thursday     70
Tuesday      50
Wednesday    60
dtype: int64

In [138]:
coffee.groupby(coffee['Units Sold'])['Day'].count().reset_index().sort_values('Units Sold')

Units Sold  Day
0          15    1
1          20    1
2          25    2
3          30    2
4          35    4
5          40    1
6          45    3

# Advanced Functionality
- .shift()
- .rank()
- .rolling()
- .cumsum()

In [140]:
coffee['Revenue'] = coffee['Units Sold']*coffee['Price']

In [141]:
coffee.head()

Day Coffee Type  Units Sold  Price  Revenue
0     Monday    Espresso          25      5      125
1     Monday       Latte          15      3       45
2    Tuesday    Espresso          30      5      150
3    Tuesday       Latte          20      3       60
4  Wednesday    Espresso          35      5      175

### 1 `.shift()`
The shift(2) method shifts the values in the Revenue column down by 2 rows. This means that each value in the Revenue column will appear two rows below its original position, and the top two rows will be filled with NaN (Not a Number) since there are no previous values to fill them.

In [142]:
coffee['Yesterday Revenue'] = coffee['Revenue'].shift(2)
coffee

Day Coffee Type  Units Sold  Price  Revenue  Yesterday Revenue
0      Monday    Espresso          25      5      125                NaN
1      Monday       Latte          15      3       45                NaN
2     Tuesday    Espresso          30      5      150              125.0
3     Tuesday       Latte          20      3       60               45.0
4   Wednesday    Espresso          35      5      175              150.0
5   Wednesday       Latte          25      3       75               60.0
6    Thursday    Espresso          40      5      200              175.0
7    Thursday       Latte          30      3       90               75.0
8      Friday    Espresso          45      5      225              200.0
9      Friday       Latte          35      3      105               90.0
10   Saturday    Espresso          45      5      225              225.0
11   Saturday       Latte          35      3      105              105.0
12     Sunday    Espresso          45      5      225              225.0
13     Sunday       Latte          35      3      105              105.0

For Tuesday (first entry), Yesterday Revenue is taken from Monday's Espresso revenue (125).

For Tuesday (second entry), it takes from Monday's Latte revenue (45).

This pattern continues for each subsequent row, where each entry in Yesterday Revenue references the revenue from two days prior.

In [145]:
coffee['Per Change'] = coffee['Revenue']/coffee['Yesterday Revenue']*100
coffee

Day Coffee Type  Units Sold  Price  Revenue  Yesterday Revenue  \
0      Monday    Espresso          25      5      125                NaN   
1      Monday       Latte          15      3       45                NaN   
2     Tuesday    Espresso          30      5      150              125.0   
3     Tuesday       Latte          20      3       60               45.0   
4   Wednesday    Espresso          35      5      175              150.0   
5   Wednesday       Latte          25      3       75               60.0   
6    Thursday    Espresso          40      5      200              175.0   
7    Thursday       Latte          30      3       90               75.0   
8      Friday    Espresso          45      5      225              200.0   
9      Friday       Latte          35      3      105               90.0   
10   Saturday    Espresso          45      5      225              225.0   
11   Saturday       Latte          35      3      105              105.0   
12     Sunday    Espresso          45      5      225              225.0   
13     Sunday       Latte          35      3      105              105.0   

    Per Change  
0          NaN  
1          NaN  
2   120.000000  
3   133.333333  
4   116.666667  
5   125.000000  
6   114.285714  
7   120.000000  
8   112.500000  
9   116.666667  
10  100.000000  
11  100.000000  
12  100.000000  
13  100.000000

### 2 `.rank()`
>**Below example**\
>Calculating Revenue Rank: The Revenue Rank column is created by applying `.rank(ascending=False)`, which ranks the revenues in descending order (higher revenues receive lower rank numbers)

In [146]:
# Adding Revenue Rank column using rank()
coffee['Revenue Rank'] = coffee['Revenue'].rank(ascending=False)
coffee

Day Coffee Type  Units Sold  Price  Revenue  Yesterday Revenue  \
0      Monday    Espresso          25      5      125                NaN   
1      Monday       Latte          15      3       45                NaN   
2     Tuesday    Espresso          30      5      150              125.0   
3     Tuesday       Latte          20      3       60               45.0   
4   Wednesday    Espresso          35      5      175              150.0   
5   Wednesday       Latte          25      3       75               60.0   
6    Thursday    Espresso          40      5      200              175.0   
7    Thursday       Latte          30      3       90               75.0   
8      Friday    Espresso          45      5      225              200.0   
9      Friday       Latte          35      3      105               90.0   
10   Saturday    Espresso          45      5      225              225.0   
11   Saturday       Latte          35      3      105              105.0   
12     Sunday    Espresso          45      5      225              225.0   
13     Sunday       Latte          35      3      105              105.0   

    Per Change  Revenue Rank  
0          NaN           7.0  
1          NaN          14.0  
2   120.000000           6.0  
3   133.333333          13.0  
4   116.666667           5.0  
5   125.000000          12.0  
6   114.285714           4.0  
7   120.000000          11.0  
8   112.500000           2.0  
9   116.666667           9.0  
10  100.000000           2.0  
11  100.000000           9.0  
12  100.000000           2.0  
13  100.000000           9.0

In [147]:
# Adding Revenue Rank column using rank()
coffee['Revenue Rank'] = coffee['Revenue Rank'].rank(ascending=True)
coffee

Day Coffee Type  Units Sold  Price  Revenue  Yesterday Revenue  \
0      Monday    Espresso          25      5      125                NaN   
1      Monday       Latte          15      3       45                NaN   
2     Tuesday    Espresso          30      5      150              125.0   
3     Tuesday       Latte          20      3       60               45.0   
4   Wednesday    Espresso          35      5      175              150.0   
5   Wednesday       Latte          25      3       75               60.0   
6    Thursday    Espresso          40      5      200              175.0   
7    Thursday       Latte          30      3       90               75.0   
8      Friday    Espresso          45      5      225              200.0   
9      Friday       Latte          35      3      105               90.0   
10   Saturday    Espresso          45      5      225              225.0   
11   Saturday       Latte          35      3      105              105.0   
12     Sunday    Espresso          45      5      225              225.0   
13     Sunday       Latte          35      3      105              105.0   

    Per Change  Revenue Rank  
0          NaN           7.0  
1          NaN          14.0  
2   120.000000           6.0  
3   133.333333          13.0  
4   116.666667           5.0  
5   125.000000          12.0  
6   114.285714           4.0  
7   120.000000          11.0  
8   112.500000           2.0  
9   116.666667           9.0  
10  100.000000           2.0  
11  100.000000           9.0  
12  100.000000           2.0  
13  100.000000           9.0

### `.cumsum()`
> **Below example:**\
Selects all columns with float data types from the 'coffee' DataFrame,\
calculates the cumulative sum for each of these columns,\
and resets the index to return a new DataFrame with a default integer index.


In [148]:
coffee.select_dtypes('float').cumsum().reset_index()

index  Yesterday Revenue   Per Change  Revenue Rank
0       0                NaN          NaN           7.0
1       1                NaN          NaN          21.0
2       2              125.0   120.000000          27.0
3       3              170.0   253.333333          40.0
4       4              320.0   370.000000          45.0
5       5              380.0   495.000000          57.0
6       6              555.0   609.285714          61.0
7       7              630.0   729.285714          72.0
8       8              830.0   841.785714          74.0
9       9              920.0   958.452381          83.0
10     10             1145.0  1058.452381          85.0
11     11             1250.0  1158.452381          94.0
12     12             1475.0  1258.452381          96.0
13     13             1580.0  1358.452381         105.0

In [150]:
coffee['Cumulative Sum'] = coffee["Revenue"].cumsum()
coffee

Day Coffee Type  Units Sold  Price  Revenue  Yesterday Revenue  \
0      Monday    Espresso          25      5      125                NaN   
1      Monday       Latte          15      3       45                NaN   
2     Tuesday    Espresso          30      5      150              125.0   
3     Tuesday       Latte          20      3       60               45.0   
4   Wednesday    Espresso          35      5      175              150.0   
5   Wednesday       Latte          25      3       75               60.0   
6    Thursday    Espresso          40      5      200              175.0   
7    Thursday       Latte          30      3       90               75.0   
8      Friday    Espresso          45      5      225              200.0   
9      Friday       Latte          35      3      105               90.0   
10   Saturday    Espresso          45      5      225              225.0   
11   Saturday       Latte          35      3      105              105.0   
12     Sunday    Espresso          45      5      225              225.0   
13     Sunday       Latte          35      3      105              105.0   

    Per Change  Revenue Rank  Cumulative Sum  
0          NaN           7.0             125  
1          NaN          14.0             170  
2   120.000000           6.0             320  
3   133.333333          13.0             380  
4   116.666667           5.0             555  
5   125.000000          12.0             630  
6   114.285714           4.0             830  
7   120.000000          11.0             920  
8   112.500000           2.0            1145  
9   116.666667           9.0            1250  
10  100.000000           2.0            1475  
11  100.000000           9.0            1580  
12  100.000000           2.0            1805  
13  100.000000           9.0            1910

In [154]:
latte = coffee[coffee['Coffee Type'] == " Latte"].copy()
latte

Day Coffee Type  Units Sold  Price  Revenue  Yesterday Revenue  \
1      Monday       Latte          15      3       45                NaN   
3     Tuesday       Latte          20      3       60               45.0   
5   Wednesday       Latte          25      3       75               60.0   
7    Thursday       Latte          30      3       90               75.0   
9      Friday       Latte          35      3      105               90.0   
11   Saturday       Latte          35      3      105              105.0   
13     Sunday       Latte          35      3      105              105.0   

    Per Change  Revenue Rank  Cumulative Sum  
1          NaN          14.0             170  
3   133.333333          13.0             380  
5   125.000000          12.0             630  
7   120.000000          11.0             920  
9   116.666667           9.0            1250  
11  100.000000           9.0            1580  
13  100.000000           9.0            1910

#### cumulative sum over three days

In [155]:
latte['3day'] = latte['Units Sold'].rolling(3).sum()
latte

Day Coffee Type  Units Sold  Price  Revenue  Yesterday Revenue  \
1      Monday       Latte          15      3       45                NaN   
3     Tuesday       Latte          20      3       60               45.0   
5   Wednesday       Latte          25      3       75               60.0   
7    Thursday       Latte          30      3       90               75.0   
9      Friday       Latte          35      3      105               90.0   
11   Saturday       Latte          35      3      105              105.0   
13     Sunday       Latte          35      3      105              105.0   

    Per Change  Revenue Rank  Cumulative Sum   3day  
1          NaN          14.0             170    NaN  
3   133.333333          13.0             380    NaN  
5   125.000000          12.0             630   60.0  
7   120.000000          11.0             920   75.0  
9   116.666667           9.0            1250   90.0  
11  100.000000           9.0            1580  100.0  
13  100.000000           9.0            1910  105.0

# New Functionality

In [156]:
coffee_numpy = pd.read_csv('../data/coffee.csv')
coffee_arrow = pd.read_csv('../data/coffee.csv', engine='pyarrow', dtype_backend='pyarrow')

In [157]:
coffee_numpy.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Day          14 non-null     object
 1   Coffee Type  14 non-null     object
 2   Units Sold   14 non-null     int64 
 3   Price        14 non-null     int64 
dtypes: int64(2), object(2)
memory usage: 580.0+ bytes


In [158]:
coffee_arrow.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 14 entries, 0 to 13
Data columns (total 4 columns):
 #   Column       Non-Null Count  Dtype          
---  ------       --------------  -----          
 0   Day          14 non-null     string[pyarrow]
 1   Coffee Type  14 non-null     string[pyarrow]
 2   Units Sold   14 non-null     int64[pyarrow] 
 3   Price        14 non-null     int64[pyarrow] 
dtypes: int64[pyarrow](2), string[pyarrow](2)
memory usage: 673.0 bytes
